src/node_api.cc

```diff
@@ -117,16 +117,23 @@ struct napi_env__ {
   CHECK_TO_TYPE((env), Boolean, (context), (result), (src), \
     napi_boolean_expected)
 
+// n-api defines NAPI_AUTO_LENGHTH as the indicator that a string
+// is null terminated. For V8 the equivalent is -1. The assert
+// validates that our cast of NAPI_AUTO_LENGTH results in -1 as
+// needed by V8.
 #define CHECK_NEW_FROM_UTF8_LEN(env, result, str, len)                   \
   do {                                                                   \
+    static_assert(static_cast<int>(NAPI_AUTO_LENGTH) == -1,              \
+                  "Casting NAPI_AUTO_LENGTH to int must result in -1");  \
     auto str_maybe = v8::String::NewFromUtf8(                            \
-        (env)->isolate, (str), v8::NewStringType::kInternalized, (len)); \
+        (env)->isolate, (str), v8::NewStringType::kInternalized,         \
+        static_cast<int>(len));                                          \
```

Comment:
@mhdawson The issue remains that this `static_cast` is invalid if `len > INT_MAX && len != NAPI_AUTO_LENGTH`, there needs to code that checks for that condition and errors out if it is met … if necessary, I *guess* that can be done in a subsequent PR, but it’s a bug that ought to be fixed.

Reply:
@addaleax thanks for the clarification, will look at that.   Since we want to have the breaking changes in before 8.6 and we are almost out of time I may do it in a follow up PR, but I'll see on Monday.

Reply:
I think this check may do the trick

Reply:
I do think you need to check for `NAPI_AUTO_LENGTH` because that value is usually larger than `INT_MAX` :) But yes, something like that sounds good

Reply:
@addaleax agreed I also need to check for NAPI_AUTO_LENGTH.  I'm going to land this PR to make sure we don't fix the Tuesday cutoff, and I'll submit a separate one to fix the bug caused by the cast on Monday.

Reply:
To be clear, the current check does not have well-defined behavior.  Compiler A might do the right thing, compiler B might do something completely different.  IOW, this is not release-ready code.